core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -263,9 +263,15 @@ private[spark] class ExecutorAllocationManager(
    */
   private def maxNumExecutorsNeeded(): Int = {
     val numRunningOrPendingTasks = listener.totalPendingTasks + listener.totalRunningTasks
-    math.ceil(numRunningOrPendingTasks * executorAllocationRatio /
-              tasksPerExecutorForFullParallelism)
-      .toInt
+    val maxNeeded = math.ceil(numRunningOrPendingTasks * executorAllocationRatio /
+      tasksPerExecutorForFullParallelism).toInt
+    if (listener.pendingSpeculativeTasks > 0 && tasksPerExecutorForFullParallelism > 1) {
```

Comment:
Why do we need to verify `tasksPerExecutorForFullParallelism ` here ?

Reply:
Because if we are using 1 task per executor, we don't need to allocate an extra executor for locality requirements.